src/coprocessor/codec/mysql/time.rs

```diff
@@ -30,6 +30,192 @@ const ZERO_DATE_STR: &'static str = "0000-00-00";
+        }
+    }
+
+    fn name_abbr(&self) -> &'static str {
+        match *self {
+            Weekday::Mon => "Mon",
+            Weekday::Tue => "Tue",
+            Weekday::Wed => "Wed",
+            Weekday::Thu => "Thu",
+            Weekday::Fri => "Fri",
+            Weekday::Sat => "Sat",
+            Weekday::Sun => "Sun",
+        }
+    }
+}
+
+pub trait DateTimeExtension {
+    fn days(&self) -> i32;
+    fn calc_year_week(
+        &self,
+        monday_first: bool,
+        zero_week: bool,
+        first_weekday: bool,
+    ) -> (i32, i32);
+    fn week(&self, monday_first: bool, zero_week: bool, first_weekday: bool) -> i32;
+}
+
+impl<Tz: TimeZone> DateTimeExtension for DateTime<Tz> {
+    /// returns the day of year starting from 1.
+    fn days(&self) -> i32 {
+        if self.month() == 0 || self.day() == 0 {
+            (0)
+        } else {
+            self.ordinal() as i32
+        }
+    }
+
+    /// returns the week of year.
+    /// when monday_first == true, Monday is considered as the first day in the week,
+    ///         otherwise Sunday.
+    /// when zero_week == true, week is from 0 to 53, otherwise from 1 to 53.
+    /// when first_weekday == true, the week that contains the first 'first-day-of-week' is week 1,
+    ///         otherwise weeks are numbered according to ISO 8601:1988.
+    fn calc_year_week(
+        &self,
+        monday_first: bool,
+        zero_week: bool,
+        first_weekday: bool,
+    ) -> (i32, i32) {
+        let week: i32;
+        let mut year = self.year();
+
+        let daynr = calc_day_number(self.year(), self.month() as i32, self.day() as i32);
+        let mut first_daynr = calc_day_number(self.year(), 1, 1);
+        let mut weekday = calc_weekday(first_daynr, !monday_first);
+        let mut week_year = !zero_week;
+        let mut days: i32;
+
+        if self.month() == 1 && (self.day() as i32) <= 7 - weekday {
+            if !week_year && ((first_weekday && weekday != 0) || (!first_weekday && weekday >= 4)) {
+                return (year, 0);
+            }
+            week_year = true;
+            year -= 1;
+            days = calc_days_in_year(year);
+            first_daynr -= days;
+            weekday = (weekday + 53 * 7 - days) % 7;
+        }
+
+        if (first_weekday && weekday != 0) || (!first_weekday && weekday >= 4) {
+            days = daynr - (first_daynr + 7 - weekday);
+        } else {
+            days = daynr - (first_daynr - weekday);
+        }
+
+        if week_year && days >= 52 * 7 {
+            weekday = (weekday + calc_days_in_year(year)) % 7;
+            if (!first_weekday && weekday < 4) || (first_weekday && weekday == 0) {
+                year += 1;
+                return (year, 1);
+            }
+        }
+        week = days / 7 + 1;
+        (year, week)
+    }
+
+    fn week(&self, monday_first: bool, zero_week: bool, first_weekday: bool) -> i32 {
+        if self.month() == 0 || self.day() == 0 {
```

Comment:
in TiDB's date implementation, month() can be 0. Although we do not support that in TiKV, we may support it in future so that I would like to keep it here.

Reply:
Keep it to be compatible with mysql.